src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/DbContextTransactionExtensions.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Threading.Tasks;
+using System.Transactions;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using Microsoft.EntityFrameworkCore;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+
+/// <summary>
+/// These extension methods provide a convenient way to ensure a transaction is
+/// created for a given unit of work.
+/// </summary>
+public static class DbContextTransactionExtensions
+{
+    /// <summary>
+    /// This method ensures that a transaction is available for the given unit of work and
+    /// is shared between any DbContexts in use during the unit of work.
+    /// 
+    /// If an existing transaction already exists, the unit of work will use the
+    /// existing transaction rather than this method creating a new one.
+    ///
+    /// If an uncaught exception is thrown during the unit of work, this will result in
+    /// a rollback, unless caught and suppressed by any parent unit of work that has its
+    /// own transaction open.
+    ///
+    /// Note that if you have a unit of work that requires interaction with multiple
+    /// DbContexts and any use the EnableRetryOnFailure() strategy, you should use any
+    /// one of the DbContexts that use this strategy to call "RequireTransaction" on,
+    /// and thereafter any other DbContexts that also use the EnableRetryOnFailure()
+    /// strategy will be functional. Not doing this will result in
+    /// InvalidOperationExceptions when those DbContexts are used. 
+    /// </summary>
+    public static async Task<TResult> RequireTransaction<TDbContext, TResult>(
```

Comment:
Just my 2 cents --> I think `RequireTransaction` is more clear as to what it is doing. I think `RunTransaction` could be interpreted as always opening up a new transaction when called. But we're actually, by default, only opening a new transaction if there isn't already an ambient transaction in scope. `RunTransaction` would also suggest that the transaction is complete after the unit of work within scope is complete - but if there is a root (ambient) transaction in scope then the transaction isn't actually complete until the outer scope calls `transactionScope.Complete`.

If we choose to add an an optional parameter to change the `TransactionScopeOption`, then we may have to consider another name. Because `TransactionScopeOption.Suppress` causes the unit of work to take part under **NO** transaction, even if there is an ambient transaction present. In this case, I'd suggest both `RequireTransaction` and `RunTransaction` are both misleading. I can't think of a good generic name that would cover all scenarios though. So I'd suggest maybe we don't create an optional parameter, and instead create additional extension methods when the scenario arises whereby we need to use the other `TransactionScopeOptions`? So keep `RequireTransaction` for `TransactionScopeOption.Required`, create a new one (only when we need to) for `TransactionScopeOption.RequiresNew` called `RequiresNewTransaction`, and create a new one (only when we need to) for `TransactionScopeOption.Suppress` called `SuppressTransaction`.

Reply:
Yeah that was along my line of thinking too as it goes @jack-hive.

Reply:
Cool, seems like a fair analysis. To be honest, I hadn't really appreciated what all the different scopes did!

It might get a little trickier if we start using the different types of scopes. There's a fair bit of nuance and complexity to this stuff so we'd want to avoid exposing too many footguns for people to trip up on.